### ISSUE TITLE
Dark text on a dark background has a light border on Linux with AA off

### ISSUE THREAD

**Issue by QuantumBadger** [QuantumBadger/Speedy2D#34]
See the following image for an example of the issue:

https://raw.githubusercontent.com/QuantumBadger/Speedy2D/81679866f7448ea8276412995f98c2cda1e2dc7a/test/assets/expected_images/test_basic_text_black_background.png

The text colour and background are set to black, but a thin white border is visible. This was discovered on Linux, with antialiasing turned off.

**Comment by QuantumBadger** [comment]
Right now I can only reproduce this issue in the unit tests. In an actual application it seems to be fine.

**Comment by matklad** [comment]
Hitting this!

Curiously, it is _not_ white background, it is actually transparent!

Here's a zoomed-in screenshot of my application on top of [two crabs](https://en.wikipedia.org/wiki/Two_Crabs).

![Screenshot_20230318_194506](https://user-images.githubusercontent.com/1711539/226134104-7793d18e-0b95-403a-a9ee-4815ae980dbb.png)

**Comment by matklad** [comment]
I _think_ what happens here is that the _alpha_ itself is blended. The followind diff uses "premultiplied alpha" trick, which seems to fix the issue for me:

```diff
diff --git a/src/font_cache.rs b/src/font_cache.rs
index 97beb05..9b5dd6e 100644
--- a/src/font_cache.rs
+++ b/src/font_cache.rs
@@ -525,9 +525,9 @@ impl BitmapRGBA
     {
         glyph.draw(|x, y, alpha| {
             let start = (4 * (self.size.x * y + x)) as usize;
-            self.data[start] = 255;
-            self.data[start + 1] = 255;
-            self.data[start + 2] = 255;
+            self.data[start] = (alpha * 255.0).round() as u8;
+            self.data[start + 1] = (alpha * 255.0).round() as u8;
+            self.data[start + 2] = (alpha * 255.0).round() as u8;
             self.data[start + 3] = (alpha * 255.0).round() as u8;
         })
     }
diff --git a/src/glbackend.rs b/src/glbackend.rs
index 22c1917..07bc7d7 100644
--- a/src/glbackend.rs
+++ b/src/glbackend.rs
@@ -56,6 +56,7 @@ pub mod constants
 
     pub const GL_SCISSOR_TEST: GLenum = glow::SCISSOR_TEST;
 
+    pub const GL_ONE: GLenum = glow::ONE;
     pub const GL_SRC_ALPHA: GLenum = glow::SRC_ALPHA;
     pub const GL_ONE_MINUS_SRC_ALPHA: GLenum = glow::ONE_MINUS_SRC_ALPHA;
 
diff --git a/src/glwrapper.rs b/src/glwrapper.rs
index dde5c29..0eb476e 100644
--- a/src/glwrapper.rs
+++ b/src/glwrapper.rs
@@ -1032,7 +1032,7 @@ impl GLContextManager
             GLBlendEnabled::Enabled(mode) => match mode {
                 GLBlendMode::OneMinusSrcAlpha => self.with_gl_backend(|backend| unsafe {
                     backend.gl_enable(GL_BLEND);
-                    backend.gl_blend_func(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+                    backend.gl_blend_func(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
                 })
             },
 ```

(not that I have any idea of what I am doing here)

**Comment by QuantumBadger** [comment]
Thank you for the pointers @matklad, that makes sense! I've spent a while looking into this and I think we can achieve the same thing using `glBlendFuncSeparate()`, which avoids the need to premultiply everything (which would include user textures, shapes, etc).

**Comment by QuantumBadger** [comment]
Now fixed in v1.12.0, thanks again @matklad!

**Comment by matklad** [comment]
Yup, tested 1.12.0, works like a charm, thanks for a quick release!
